coco/scheduler.py

```diff
@@ -3,3 +3,139 @@
+        ----------
+        endpoints : dict of Endpoint
+            All endpoints configured on coco.
+        host : str
+            Hostname for coco.
+        port : int
+            Port for coco.
+        """
+        self.host, self.port = host, port
+        # find endpoints to schedule
+        self._gen_timers(endpoints)
+
+    async def start(self):
+        """Start the scheduler (async)."""
+        self.tasks = []
+        for timer in self.timers:
+            logger.info(f"Setting timer '{timer.name}' every {timer.period} s.")
+            task = asyncio.create_task(timer.run())
+            self.tasks.append(task)
+        await asyncio.gather(*self.tasks)
+
+    def stop(self):
+        """Stop the scheduler. Cancels all timer tasks."""
+        for task in self.tasks:
+            task.cancel()
+
+    def _gen_timers(self, endpoints):
+        if len(self.timers) > 0:
+            raise Exception("Timers already exist.")
+
+        for edpt in endpoints.values():
+            if edpt.schedule is not None:
+                try:
+                    period = edpt.schedule["period"]
+                except KeyError:
+                    raise ValueError(
+                        f"Endpoint '{edpt.name}' schedule block must include 'period'."
+                    )
+                period = str2total_seconds(period)
+                if period is None or period == 0:
+                    raise ValueError(
+                        f"Could not parse 'period' parameter for endpoint {edpt.name}"
+                    )
+                timer = EndpointTimer(period, edpt, self.host, self.port)
+                self.timers.append(timer)
+
+
+class Timer(object):
+    """Asynchronous timer."""
+
+    def __init__(self, name, period):
+        self.name = name
+        self.period = period
+        self._last_t = time()
+        self._start_t = self._last_t
+
+    async def run(self):
+        """Start the timer (async)."""
+        while True:
+            t = self._wait_time()
+            if t > 0:
+                try:
+                    await asyncio.sleep(t)
+                except asyncio.CancelledError:
+                    logger.debug(f"Cancelled timer '{self.name}''.")
+                    break
+            self._last_t = time()
+            await self._call()
+
+    async def _call(self):
+        """Override this method with whatever the timer does."""
+        pass
+
+    def _wait_time(self):
+        """Time to wait until next execution."""
+        return self.period - (time() - self._last_t)
+
+
+class EndpointTimer(Timer):
+    """Timer that calls a coco endopint."""
+
+    def __init__(self, period, endpoint, host, port):
+        self.endpoint = endpoint
+        self.host, self.port = host, port
+        super().__init__(endpoint.name, period)
+
+    async def _call(self):
+        logger.debug(f"{self.name}: {time() - self._start_t}")
+        # Use stored endpoint data
+        data = copy(self.endpoint.values)
```

Comment:
`Endpoint.values` is the configured expected types a request to this endpoint will be checked for before getting forwarded, it does not include the actual data to attach to the request.

As described above, I think the scheduler should (on start) make sure that none of its endpoint has this set, because it wouldn't know what values to use and therefor any call would fail.